### DISCUSSION TITLE
How to train with cosine annealing?

### DISCUSSION THREAD

**Discussion by ZQ-Dev8** [axolotl-ai-cloud/axolotl#1450] [Q&A]
Is it possible to train a model with axolotl using some form of [cosine annealing with warm restarts](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.CosineAnnealingWarmRestarts.html)? A hacky manual way would be to train individual epochs with the default cosine scheduler + warmup steps + dataset shuffle, but I seem to remember this was once supported out-of-the-box?

**Comment by NanoCode012** [comment]
Can you check if this function `get_cosine_schedule_with_min_lr` within the code does what you want?

**Reply by ZQ-Dev8** [reply]
@NanoCode012 unless I'm missing something, it does not appear to without some level of modification. What it needs to do is essentially an entire iteration of a cosine schedule, except across a single epoch. So for each epoch during training, there are warmup steps up to a max_lr, with cosine decay down to a min_lr. I've also seen variations of this where the max_lr is reduced per epoch.

**Comment by NanoCode012** [comment]
For any future readers, we do support `lr_scheduler: one_cycle` to use a cosine annealing. I think this uses the `warmup_steps` (or ratio if set) to determine the percentage of cycle spent increasing lr (see `pct_start` in scheduler's definition)
